### PR TITLE
[reward, perf, tests] feat: batch PickScore inference

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

PickScore requests were previously processed one at a time, even though the reward loop submits samples concurrently. This change collects each ready burst into a bounded micro-batch of up to 16 requests and deduplicates repeated prompts before text encoding.

The original score contract is preserved: results are returned in request order, each prompt remains paired with its image, invalid image conversion fails only that request, and public configuration and return values are unchanged.

This does not duplicate an existing contribution. Searches performed before submission:

- https://github.com/verl-project/verl-omni/pulls?q=is%3Apr+is%3Aopen+PickScore+batching
- https://github.com/verl-project/verl-omni/issues?q=is%3Aissue+is%3Aopen+PickScore

AI assistance was used. The human submitter has completed the repository-required line-by-line review.

### Checklist Before Starting

- [x] Search for similar PRs. Query links are included above.
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Test

CPU unit tests:

```text
python -m pytest -q --asyncio-mode=auto \
  tests/utils/reward_score/test_pickscore_reward_on_cpu.py \
  tests/utils/reward_score/test_choice_reward_on_cpu.py \
  tests/utils/reward_score/test_latent_http_scorer_client_on_cpu.py \
  tests/utils/reward_score/test_reward_utils_on_cpu.py

22 passed in 2.24s
```

Repository checks:

```text
SKIP=autogen-trainer-cfg pre-commit run --all-files --show-diff-on-failure

All executed hooks passed. autogen-trainer-cfg was skipped because the local
validation environment did not contain the repository-pinned verl package;
this PR does not change trainer configuration or generated YAML.
```

Real-model GPU validation on one NVIDIA A800 80GB PCIe, PyTorch 2.6.0+cu124, Transformers 5.9.0, float32, 512x512 images, batch size 16, five measured repeats after warm-up:

| Metric | main | this PR |
|---|---:|---:|
| Median latency per 16 requests | 0.7853 s | 0.4556 s |
| Median throughput | 20.37 images/s | 35.12 images/s |
| Peak allocated memory | 3.76 GiB | 3.99 GiB |
| Peak reserved memory | 4.156 GiB | 4.156 GiB |

- Median speedup: **1.724x**
- Latency reduction: **41.98%**
- Throughput increase: **72.37%**
- Maximum score absolute difference: **1.0133e-6**
- `torch.allclose(rtol=1e-4, atol=1e-5)`: passed
- Mechanism checks: main used 16 singleton image/text calls; this PR used one image batch of 16 and one unique prompt encoding per burst.

The A800 run used production source SHA256 `b87ba51a7ea00a494b1b735eac7f2dd63e5fa24a4aebf7b410304e898bf77591`, which matches the production file in this commit.

### API and Usage Example

No API or configuration changes. Existing callers continue to use `compute_score_pickscore` unchanged.

### Design & Code Changes

- collect currently queued requests after one event-loop yield, capped at 16;
- preprocess and infer valid images as one batch while isolating conversion failures;
- encode each distinct prompt once, then expand embeddings back to request order;
- add CPU tests for prompt deduplication, ordering, batch capping, and invalid-image isolation.

### Checklist Before Submitting

- [x] Read the contribution guide.
- [x] Run relevant formatting, linting, type, structure, license, and compile checks.
- [x] Add CPU tests collected by the existing `*_on_cpu.py` CI convention.
- [x] Documentation is not applicable because no API, configuration, model-support, or workflow contract changes.
- [x] Human submitter has reviewed every changed line before marking the PR ready.

### CODE REVIEW

**Comment by zhtmike** [comment]
we need a e2e reward curve for this change. thx

**Comment by WenzheWang** [comment]
Added the requested end-to-end reward curve on a real PickScore scorer.

### Setup

- Source: `main` at `e86abf821ca4282f7abc7bcf94fe5f5251673e16` vs this PR at `fe1ad96d21f8ac3e2d2826617335b307322b90aa`
- Hardware: 1x NVIDIA A800 80GB
- Runtime: PyTorch 2.11.0+cu129, vLLM 0.24.0+cu129, vLLM-Omni 0.24.0, Transformers 5.9.0
- Workload: fully offline tiny random SD3, real cached PickScore v1 / CLIP-ViT-H-14 processor
- Training: 5 steps, seed 42, train batch 6, `n=2` (12 images/step), 3 unique reward prompts
- Execution: paired sequential runs with identical model, data, seed, and training configuration

### Reward curve

```mermaid
xychart-beta
    title "PR #346 E2E PickScore reward curve"
    x-axis "Training step" [1, 2, 3, 4, 5]
    y-axis "Mean reward" 0.753 --> 0.769
    line [0.75901419, 0.76825935, 0.76587826, 0.75587112, 0.75349754]
    line [0.75902206, 0.76835012, 0.76579505, 0.75570124, 0.75332642]
```

Series declaration order: `main`, then PR #346. The repeated-`main` control is included in the table below.

| Step | `main` | PR #346 | abs delta | `main` repeat control |
|---:|---:|---:|---:|---:|
| 1 | 0.75901419 | 0.75902206 | 7.87e-06 | 0.75902170 |
| 2 | 0.76825935 | 0.76835012 | 9.08e-05 | 0.76832294 |
| 3 | 0.76587826 | 0.76579505 | 8.32e-05 | 0.76568437 |
| 4 | 0.75587112 | 0.75570124 | 1.70e-04 | 0.75571531 |
| 5 | 0.75349754 | 0.75332642 | 1.71e-04 | 0.75339127 |

All 15 points are finite. The PR curve tracks `main` closely:

- `main` vs PR: max absolute delta `1.711e-4`, mean absolute delta `1.046e-4`, Pearson correlation `0.999936`
- `main` vs repeated `main` control: max absolute delta `1.939e-4`, mean absolute delta `1.054e-4`

The PR delta is within the observed same-commit run-to-run jitter, so this short conformance run does not show a reward-curve or precision regression.

For full-step timing (excluding the first-step model/scorer cold start), steps 2-5 averaged `5.106s` on `main` and `4.920s` on the PR: `1.038x` speedup / `3.65%` lower step time. This complements the scorer-only A800 benchmark already in the PR description (`1.724x` scorer throughput).

This is a tiny-model integration/conformance check, not a model-quality claim. It covers the actual rollout -> PickScore reward -> advantage -> actor update path requested here.

Could you please re-review? If the result looks sufficient, please also trigger the label-gated GPU CI (`ci-e2e-diffusion` or `ready-for-ci`, as appropriate).

**Comment by zhtmike** [comment]
Possible to show a jpeg directly in the pr description?

**Comment by zhtmike** [comment]
And pls run with bagel with pickscore e2e script (check repo we have it), check the gpu memory usage and potentially oom ( i majorly consider this)

**Comment by WenzheWang** [comment]
@zhtmike Requested BAGEL + PickScore E2E run completed on PR head `fe1ad96d`: 4x A800 80GB, batch 48, `n=16`, 768 images, 15 diffusion steps, `global_step=1`, exit `0:0`. Peak memory was 49.75/45.00/45.16/45.04 GiB (minimum headroom 30.25 GiB); no OOM or fatal error. Mean reward 0.8205854297 vs raw PickScore 0.8205854446 (delta 1.50e-8); loss, grad norm, and PPO KL were finite. Please re-review.

**Comment by WenzheWang** [comment]
Peak memory chart for the BAGEL + PickScore E2E run (bars: used; line: 80 GiB capacity):

```mermaid
xychart-beta
    title "Peak A800 memory (GiB)"
    x-axis [GPU0, GPU1, GPU2, GPU3]
    y-axis "GiB" 0 --> 80
    bar [49.75, 45.00, 45.16, 45.04]
    line [80, 80, 80, 80]
```

**Comment by WenzheWang** [comment]
image

**Comment by zhtmike** [comment]
looks good

**Review by zhtmike — APPROVED** [review]
